ui/src/app/utils/cluster-initializer.tsx

```diff
@@ -0,0 +1,301 @@
+ *  - Cluster initialization can take up to 10 minutes, which is beyond the scope of a single
+ *    App Engine server-side request timeout. To reliably implement this control on the server side
+ *    would likely require new database persistence, tasks queues, and additional APIs in order to
+ *    provide the client with status updates.
+ *  - Ultimately, we might expect this type of functionality ("Get me a cluster for workspace X and
+ *    bring it to a running state") to exist as part of the Leonardo application. So rather than
+ *    build our own server-side equivalent, we adopted a client-side solution as a holdover.
+ */
+export class ClusterInitializer {
+  // Core properties for interacting with the caller and the cluster APIs.
+  private workspaceNamespace: string;
+  private onStatusUpdate: (ClusterStatus) => void;
+  private abortSignal?: AbortSignal;
+
+  // Properties to track & control the polling loop. We use a capped exponential backoff strategy
+  // and a series of "maxFoo" limits to ensure the initialization flow doesn't get out of control.
+  private currentDelay: number;
+  private maxDelay: number;
+  private overallTimeout: number;
+  private maxCreateCount: number;
+  private maxDeleteCount: number;
+  private maxResumeCount: number;
+  private maxServerErrorCount: number;
+
+  // Properties to track progress, actions taken, and errors encountered.
+  private createCount = 0;
+  private deleteCount = 0;
+  private resumeCount = 0;
+  private serverErrorCount = 0;
+  private initializeStartTime?: number;
+  // The latest cluster retrieved from getCluster. If the last getCluster call returned a NOT_FOUND
+  // response, this will be null.
+  private currentCluster?: Cluster;
+
+  // Properties to control the initialization and promise resolution flow.
+  //
+  // Tracks whether the .run() method has been called yet on this class. Each instance is only
+  // allowed to be used for one initialization flow.
+  private isInitializing = false;
+  // The resolve and reject function from the promise returned from the call to .run(). We use a
+  // deferred-style approach in this class, which allows us to provide a Promise-based API on the
+  // .run() method, but to call the resolve() or reject() method from anywhere in this class.
+  private resolve: (cluster?: Cluster | PromiseLike<Cluster>) => void;
+  private reject: (error: Error) => void;
+
+  constructor(options: ClusterInitializerOptions) {
+    this.workspaceNamespace = options.workspaceNamespace;
+    this.onStatusUpdate = options.onStatusUpdate ? options.onStatusUpdate : () => {};
+    this.abortSignal = options.abortSignal;
+    this.currentDelay = options.initialPollingDelay != null ? options.initialPollingDelay : DEFAULT_INITIAL_POLLING_DELAY;
+    this.maxDelay = options.maxPollingDelay != null ? options.maxPollingDelay : DEFAULT_MAX_POLLING_DELAY;
+    this.overallTimeout = options.overallTimeout != null ? options.overallTimeout : DEFAULT_OVERALL_TIMEOUT;
```

Comment:
FWIW, I think you can just do `this.overallTimeout = this.overallTimeout || DEFAULT_OVERALL_TIMEOUT`. Have to be careful with builtins though.

(nit): Also, I'd probably invert the condition so you're checking  `== null`, just to avoid negation wherever possible.


Finally, there's got to be some kind of builtin notion of a class/struct property with a default value.